src/sentry/workflow_engine/processors/delayed_workflow.py

```diff
@@ -264,6 +276,129 @@ def get_groups_to_fire(
+    }
+    node_ids = list(node_id_to_event_id.keys())
+    fetch_retry_policy = ConditionalRetryPolicy(should_retry_fetch, exponential_delay(1.00))
+
+    bulk_data = {}
+    for node_id_chunk in chunked(node_ids, EVENT_LIMIT):
+        bulk_results = fetch_retry_policy(lambda: nodestore.backend.get_multi(node_id_chunk))
+        bulk_data.update(bulk_results)
+
+    return {
+        node_id_to_event_id[node_id]: Event(
+            event_id=node_id_to_event_id[node_id], project_id=project_id, data=data
+        )
+        for node_id, data in bulk_data.items()
+        if data is not None
+    }
+
+
+def get_group_to_groupevent(
+    dcg_group_to_event_data: DataConditionGroupEvent,
+    group_ids: list[int],
+    event_ids: set[str],
+    occurrence_ids: set[str],
+    project_id: int,
+) -> dict[Group, GroupEvent]:
+    groups = Group.objects.filter(id__in=group_ids)
+    group_id_to_group = {group.id: group for group in groups}
+
+    bulk_event_id_to_events = bulk_fetch_events(list(event_ids), project_id)
+    bulk_occurrences = IssueOccurrence.fetch_multi(list(occurrence_ids), project_id=project_id)
+
+    bulk_occurrence_id_to_occurrence = {
+        occurrence.id: occurrence for occurrence in bulk_occurrences if occurrence
+    }
+
+    group_to_groupevent: dict[Group, GroupEvent] = {}
+    for dcg_group, instance_data in dcg_group_to_event_data.items():
+        event_id = instance_data.get("event_id")
+        occurrence_id = instance_data.get("occurrence_id")
+
+        if event_id is None:
+            continue
+
+        event = bulk_event_id_to_events.get(event_id)
+        group = group_id_to_group.get(int(dcg_group[1]))
+
+        if not group or not event:
+            continue
+
+        group_event = event.for_group(group)
+        if occurrence_id:
+            group_event.occurrence = bulk_occurrence_id_to_occurrence.get(occurrence_id)
+        group_to_groupevent[group] = group_event
+
+    return group_to_groupevent
+
+
+def fire_actions_for_groups(
+    groups_to_fire: dict[int, set[DataConditionGroup]],
+    trigger_type_to_dcg_model: dict[DataConditionHandlerType, dict[int, int]],
+    group_to_groupevent: dict[Group, GroupEvent],
+) -> None:
+    for group, group_event in group_to_groupevent.items():
+        job = WorkflowJob({"event": group_event})
+        detector = get_detector_by_event(job)
+
+        workflow_triggers: set[DataConditionGroup] = set()
+        action_filters: set[DataConditionGroup] = set()
+        for dcg in groups_to_fire[group.id]:
+            if dcg.id in trigger_type_to_dcg_model[DataConditionHandlerType.WORKFLOW_TRIGGER]:
+                workflow_triggers.add(dcg)
+            elif dcg.id in trigger_type_to_dcg_model[DataConditionHandlerType.ACTION_FILTER]:
+                action_filters.add(dcg)
+
+        filtered_actions: list[Action] = []
+
+        # process action_filters
+        filtered_actions.extend(list(filter_recently_fired_workflow_actions(action_filters, group)))
```

Comment:
are these on separate lines for typing reasons?

Reply:
no. they can be on one line